arch/arm/core/cortex_m/mpu/nxp_mpu.c

```diff
@@ -411,80 +555,55 @@ int arm_core_mpu_buffer_validate(void *addr, size_t size, int write)
  * This function provides the default configuration mechanism for the Memory
  * Protection Unit (MPU).
  */
-static void _nxp_mpu_config(void)
+static int nxp_mpu_init(struct device *arg)
 {
+	ARG_UNUSED(arg);
+
 	u32_t r_index;
 
-	__ASSERT(mpu_config.num_regions <= _get_num_regions(),
-		 "too many static MPU regions defined");
+	if (mpu_config.num_regions > _get_num_regions()) {
+		/* Attempt to configure more MPU regions than
+		 * what is supported by hardware. As this operation
+		 * may be executed during system (pre-kernel) initialization,
+		 * we want to ensure we can detect an attempt to
+		 * perform invalid configuration.
+		 */
+		__ASSERT(0,
+			"Request to configure: %u regions (supported: %u)\n",
+			mpu_config.num_regions,
+			_get_num_regions()
+		);
+		return -1;
+	}
+
 	LOG_DBG("total region count: %d", _get_num_regions());
 
-	/* Disable MPU */
-	SYSMPU->CESR &= ~SYSMPU_CESR_VLD_MASK;
-	/* Clear MPU error status */
-	SYSMPU->CESR |=  SYSMPU_CESR_SPERR_MASK;
+	arm_core_mpu_disable();
 
-	/* MPU Configuration */
+	/* Architecture-specific configuration */
+	_mpu_init();
 
-	/* Configure regions */
+	/* Program fixed regions configured at SOC definition. */
 	for (r_index = 0U; r_index < mpu_config.num_regions; r_index++) {
-		_region_init(r_index,
-			     mpu_config.mpu_regions[r_index].base,
-			     mpu_config.mpu_regions[r_index].end,
-			     mpu_config.mpu_regions[r_index].attr);
+		_region_init(r_index, &mpu_config.mpu_regions[r_index]);
 	}
 
-	/* Enable MPU */
-	SYSMPU->CESR |= SYSMPU_CESR_VLD_MASK;
+	/* Update the number of programmed MPU regions. */
+	static_regions_num = mpu_config.num_regions;
 
-#if defined(CONFIG_APPLICATION_MEMORY)
-	u32_t index, region_attr, base, size;
 
-	/* configure app data portion */
-	index = _get_region_index_by_type(THREAD_APP_DATA_REGION);
-	region_attr = _get_region_attr_by_type(THREAD_APP_DATA_REGION);
-	base = (u32_t)&__app_ram_start;
-	size = (u32_t)&__app_ram_end - (u32_t)&__app_ram_start;
+	arm_core_mpu_enable();
 
-	/* set up app data region if exists, otherwise disable */
-	if (size > 0) {
-		_region_init(index, base, ENDADDR_ROUND(base + size),
-			     region_attr);
-	} else {
-		SYSMPU->WORD[index][3] = 0;
-	}
-#endif
 	/* Make sure that all the registers are set before proceeding */
 	__DSB();
 	__ISB();
```

Comment:
Remove the barriers at the end of nxp_mpu_init().

Reply:
Done.